### PR TITLE
 Add dedicated health port for JITServer

### PR DESCRIPTION
**Pull request by mpirvu**
Kubernetes can use liveness and readiness probes to check
a pod's health. For a TCP port, the kubelet attempts to open a
socket to the container on the specified port. If a connection
can be established, the container is considered healthy.

The JITServer did not have a dedicated port for health probes;
instead, the main communication port (default 38400) could be
used for this purpose. However, this solution is not ideal
when communication between the client and server is encrypted
with OpenSSL. Starting with OpenSSL 3.0, if one of the participants
closes the connection without notifying the other party, the
OpenSSL library will log error messages like:
"A0F04FE4FE7F0000:error:0A000126:SSL routines:ssl3_read_n:unexpected eof while reading:ssl/record/rec_layer_s3.c:320"
Since the health probes are sent periodically, over time, the log
of the JITServer container could be flooded with such messages.
While the error is benign in nature, these messages are annoying
and can create concerns for end-users.

This commit adds a dedicated health port for JITServer, where
communication is never encrypted. By default, the JITServer will
open a health port and the value of this port is 38600.
The following options were added:
`-XX:JITServerHealthProbePort=NNN` # change the health port value
`-XX:-JITServerHealthProbes` # disable the dedicated health port
`-XX:+JITServerHealthProbes` # enable the dedicated health port

Issue: https://github.com/OpenLiberty/open-liberty/issues/27665

### CODE REVIEW

**Comment by mpirvu** [comment]
@dsouzai Could you please review/merge this PR? Thanks

File: runtime/compiler/runtime/Listener.cpp [hunk 1]
```diff
@@ -45,10 +45,34 @@
...
+   unsigned long errorCode = 0;
+   int errorReason = 0;
+   char errorString[256] = {0};
+
+   // Filter out SSL_R_UNEXPECTED_EOF_WHILE_READING errors that were introduced in SSL version 3
+   // These can appear if Kubernetes readiness/liveness probes are used on the TCP encrypted channel
+   if (ret <= 0)
+      {
+      sslError = (*OSSL_get_error)(ssl, ret);
+      // Get earliest error code from the thread's error queue without modifying it.
```
Comment (dsouzai) [inline runtime/compiler/runtime/Listener.cpp:61]: Is it possible for the error to not be the first one?

File: runtime/compiler/runtime/Listener.cpp [hunk 2]
```diff
@@ -45,10 +45,34 @@
...
+      errorCode = OERR_peek_error();
+      errorReason = ERR_GET_REASON(errorCode);
+      OERR_error_string_n(errorCode, errorString, sizeof(errorString));
+      if (sslError == SSL_ERROR_SSL && errorReason == SSL_R_UNEXPECTED_EOF_WHILE_READING)
+         {
+         // Take this error out of the queue so that it doesn't get printed again
```
Comment (dsouzai) [inline runtime/compiler/runtime/Listener.cpp:67]: When you say "printed again", when was it printed before?

File: runtime/compiler/runtime/Listener.cpp [hunk 3]
```diff
@@ -179,49 +261,64 @@ TR_Listener::serveRemoteCompilationRequests(BaseCompileDispatcher *compiler)
...
-               perror("Can't set option SO_RCVTIMEO on connfd socket");
-               exit(1);
-               }
-            if (setsockopt(connfd, SOL_SOCKET, SO_SNDTIMEO, &timeout, sizeof(timeout)) < 0)
-               {
-               perror("Can't set option SO_SNDTIMEO on connfd socket");
-               exit(1);
+               if (fdIndex == 0) // readiness/liveness probe socket
+                  {
+                  close(connfd);
```
Comment (dsouzai) [inline runtime/compiler/runtime/Listener.cpp:292]: Does the probe only need a connection to be accepted and not anything that the jitserver has to send back?

**Review by dsouzai — COMMENTED** [review]
Overall LGTM, just had some questions.

File: runtime/compiler/runtime/Listener.cpp [hunk 4]
```diff
@@ -45,10 +45,34 @@
...
+   int errorReason = 0;
+   char errorString[256] = {0};
+
+   // Filter out SSL_R_UNEXPECTED_EOF_WHILE_READING errors that were introduced in SSL version 3
+   // These can appear if Kubernetes readiness/liveness probes are used on the TCP encrypted channel
+   if (ret <= 0)
+      {
+      sslError = (*OSSL_get_error)(ssl, ret);
+      // Get earliest error code from the thread's error queue without modifying it.
+      errorCode = OERR_peek_error();
```
[the remark this answers is not in the document]
Comment (mpirvu) [reply runtime/compiler/runtime/Listener.cpp:62]: Both are correct. The first one is more "in your face" that we are dealing with a function pointer. Existing code seems to prefer this variant, so I will change my code to do the same for uniformity.

File: runtime/compiler/runtime/Listener.cpp (hunk 2 above)
Comment (mpirvu) [reply runtime/compiler/runtime/Listener.cpp:67]: I deleted "again". 
The error will get printed in `(*OERR_print_errors_fp)(stderr);` if we don't take it out of the queue.

File: runtime/compiler/runtime/Listener.cpp (hunk 3 above)
Comment (mpirvu) [reply runtime/compiler/runtime/Listener.cpp:292]: Kubernetes documentation says that establishing the connection is enough. No data has to be sent and my tests validated that.

**Comment by mpirvu** [comment]
@KostasTsiounis Could you please review the changes that involve OpenSSL usage, i.e. https://github.com/eclipse-openj9/openj9/pull/19107/commits/52145364666045c45fa5c70f058fdb2a94dce01c
Thanks

File: runtime/compiler/runtime/Listener.cpp [hunk 5]
```diff
@@ -45,10 +45,34 @@
...
+         errorCode = (*OERR_get_error)();
+         }
+      }
    if (TR::Options::getVerboseOption(TR_VerboseJITServer))
-       TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "%s: errno=%d", errMsg, errno);
+       TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "t=%lu %s: errno=%d sslError=%d errorString=%s",
+                                      (unsigned long)compInfo->getPersistentInfo()->getElapsedTime(),
+                                      errMsg, errnoCopy, sslError, errorString);
+   // Print the error strings for all errors that OpenSSL has recorded and empty the error queue
    (*OERR_print_errors_fp)(stderr);
```
Comment (KostasTsiounis) [inline runtime/compiler/runtime/Listener.cpp:76]: Won't this always print the errors in `stderr`? Unless you only care about what's printed on `stdout`.

Comment (KostasTsiounis) [reply runtime/compiler/runtime/Listener.cpp:76]: Yes, but you seem to be calling `ERR_peek_error()` a few lines above, which doesn't remove it from the queue. Are you calling the `ERR_get_error()` at some other point that I am missing?

Comment (mpirvu) [reply runtime/compiler/runtime/Listener.cpp:76]: The intended logic is to print all errors, except `SSL_R_UNEXPECTED_EOF_WHILE_READING`.
I use `ERR_peek_error()` to determine the error reason, and if that is `SSL_R_UNEXPECTED_EOF_WHILE_READING`, then I use `ERR_get_error()` the eliminate from the queue and prevent it from being printed.

Comment (KostasTsiounis) [reply runtime/compiler/runtime/Listener.cpp:76]: Ok, cool. I didn't see the `ERR_get_error()` in the if statement. LGTM otherwise.

**Comment by mpirvu** [comment]
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit,alinux64jit jdk17

**Comment by mpirvu** [comment]
Lots of:
```
 [ERR] can't bind server address: Address already in use
 [ERR] Failed to open health socket on port 38600
```
I guess I have to fix the tests.

**Comment by mpirvu** [comment]
jenkins test sanity xlinuxjit jdk17

File: test/functional/cmdLineTests/jitserver/jitserverScript.sh [hunk 6]
```diff
@@ -28,17 +28,20 @@ echo "start running script";
...
-# %5 is the Metrics
+# $5 is the Metrics
+# $6 is a boolean for testing the health port
 
 TEST_ROOT=$1
 TEST_JDK_BIN=$2
 JITSERVER_OPTS="$3"
 JVM_OPTS="$4"
 METRICS=$5
+HEALTHPORTTEST=$6
```
Comment (dsouzai) [inline test/functional/cmdLineTests/jitserver/jitserverScript.sh:39]: Might also need to add this to `criu/criuJitServerScript.sh`

**Comment by mpirvu** [comment]
jenkins test sanity plinuxjit,zlinuxjit,alinux64jit jdk17

**Comment by mpirvu** [comment]
Tests have been fixed. All tests pass now.

**Review by dsouzai — APPROVED** [review]

**Pull request merged**